File: 105. Construct Binary Tree from Preorder and Inorder Traversal/main.cpp
#include <unordered_map>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr) {}
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x)
        , left(left)
        , right(right) {}
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    TreeNode *build(vector<int> &preorder, int preLeft, int preRight, vector<int> &inorder, int inLeft, int inRight) {
        if (preLeft > preRight || inLeft > inRight) {
            return nullptr;
        }

        int mid = 0;

        for (int i = inLeft; i <= inRight; i++) {
            if (inorder[i] == preorder[preLeft]) {
                mid = i;
            }
        }

        TreeNode *node = new TreeNode(preorder[preLeft]);
        node->left = build(preorder, preLeft + 1, preLeft + mid - inLeft, inorder, inLeft, mid - 1);
        node->right = build(preorder, preLeft + mid - inLeft + 1, preRight, inorder, mid + 1, inRight);

        return node;
    }

public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        int preorderSize = preorder.size();
        int inorderSize = inorder.size();

        return build(preorder, 0, preorderSize - 1, inorder, 0, inorderSize - 1);
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    unordered_map<int, int> map;
    int index = 0;

    TreeNode *build(vector<int> &preorder, vector<int> &inorder, int start, int end) {
        if (start > end) {
            return nullptr;
        }

        int current = preorder[index++];
        TreeNode *node = new TreeNode(current);
        int idx = map[current];
        node->left = build(preorder, inorder, start, idx - 1);
        node->right = build(preorder, inorder, idx + 1, end);

        return node;
    }

public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        int size = inorder.size();

        for (int i = 0; i < size; i++) {
            map[inorder[i]] = i;
        }

        return build(preorder, inorder, 0, size - 1);
    }
};

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/main.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if len(preorder) == 0 or len(inorder) == 0:
            return None

        node = TreeNode(preorder[0])
        mid = inorder.index(preorder[0])
        node.left = self.buildTree(preorder[1 : mid + 1], inorder[:mid])
        node.right = self.buildTree(preorder[mid + 1 :], inorder[mid + 1 :])

        return node


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        index = 0
        map = dict()

        for i in range(len(inorder)):
            map[inorder[i]] = i

        def helper(start: int, end: int) -> Optional[TreeNode]:
            nonlocal index

            if start > end:
                return None

            current = preorder[index]
            index += 1
            node = TreeNode(current)
            idx = map[current]

            node.left = helper(start, idx - 1)
            node.right = helper(idx + 1, end)

            return node

        return helper(0, len(inorder) - 1)
